### ISSUE TITLE
Model Development

### ISSUE THREAD

**Issue by makenahillman** [makenahillman/on-time_voyager#5]

**Comment by mfirn7894** [comment]
**Predict P(delay)**

- Load cleaned dataset into Pandas DF
- Feature engineering
    - Normalize and transform metadata features
    - Aggregate upstream flight info
- Build training DF
    - Build labels based on delay threshold
    - Stratified split into train/test, oversampling minority class for training set but preserving original class balance for test set
- Train model
    - Build grid of algorithm/hyperparameter combos
    - Build preprocessing pipeline (e.g. PCA)
    - Train using k-fold CV
    - Record and save winning model using average CV score
- Evaluate
    - Compute ROC AUC on test set
    - Compute F1 on test set

**Post-Processing**

- Probability calibration
    - Bin test set scores
    - Fit monotonic curves (exp, log, linear, arctan, sigmoid, etc.)
    - Store parameters of best fit
- Ensure curves are monotonic (i.e. P(2h delay) <= P(1h delay))

**Delay Reason Decomposition**

- Find k-nearest neighbors among delayed flights
- Look at distribution of delay reasons for those neighbors
    - This may be prohibitively slow for real-time inference
        - Sample for the largest amount we can store in memory?
        - Would like to avoid building specific models for each reason

**Save Model Objects**

- Trained models (.pkl?)
- Calibration params

**Comment by taylor-boyd** [comment]
- [ ] https://github.com/makenahillman/on-time_voyager/issues/9
